clarity-ext-scripts/clarity_ext_scripts/covid/rtpcr_analysis_service.py

```diff
@@ -1,37 +1,107 @@
+        elif fam_ct <= self.FAM_THRESHOLD:
+            return COVID_RESPONSE_POSITIVE
+        else:
+            # TODO Don't know if it is actually a good idea to raise here or not...
+            raise NotImplementedError(
+                "Got CT-value for FAM: {} and CT-value for HEX: {}.".format(fam_ct, hex_ct))
 
-    def validate_control_value(self, control_type, ct_value, lower_bound, upper_bound):
-        if control_type.lower() == RT_PCR_POSITIVE_CONTROL and ct_value <= lower_bound:
-            self.has_valid_pos_control = True
-        if control_type.lower() == RT_PCR_NEGATIVE_CONTROL and ct_value >= upper_bound:
-            self.has_valid_neg_control = True
+    def _analyze_controls(self, positive_controls, negative_controls):
+        errors = []
+        for pos_control in positive_controls:
+            if self._analyze_sample(pos_control) == COVID_RESPONSE_NEGATIVE:
+                errors += PositiveControlWasNegative(
+                    "Positive control sample: {} was negative for covid-19".format(pos_control["id"]))
 
-    def is_valid(self):
-        return self.has_valid_neg_control and self.has_valid_pos_control
+        for neg_control in negative_controls:
+            if self._analyze_sample(neg_control) == COVID_RESPONSE_POSITIVE:
+                errors += NegativeControlWasPositive(
+                    "Negative control sample: {} was positive for covid-19".format(
+                        neg_control["id"]))
 
-    def analyze(self, ct_value, lower_bound, upper_bound):
+        if errors:
+            for e in errors:
+                log.error(e.message)
+            raise MultipleAnalysisErrors(errors)
+
+    def analyze_samples(self, positive_controls, negative_controls, samples):
         """
-        Returns: String with value of either 'positive', 'negative', 'failed'
+        This assumes all controls and samples are from the same plate.
+        All samples, and controls should be submitted as dict-like objects on the format:
+         {"id": "<sample id>", "FAM-CT": <value as numeric>, "HEX-Ct": <value as numric>}
+
+        The method will return a generator of objects on the format:
+         {"id": "<same as sample id above>",
+             "diagnosis_result": "<CONST COVID RESPONSE>"}
         """
-        if not self.has_valid_pos_control or not self.has_valid_neg_control:
-            raise UsageError("There are no valid control samples")
-        if ct_value <= lower_bound:
-            return 'positive'
-        elif ct_value >= upper_bound:
-            return 'negative'
-        return 'failed'
+
+        # TODO Should we validate that there are controls on the plate?
+
+        # If controls fail, fail all samples on plate.
+        try:
+            self._analyze_controls(positive_controls, negative_controls)
+        except MultipleAnalysisErrors:
+            for sample in samples:
+                yield {"id": sample["id"], "diagnosis_result": COVID_RESPONSE_FAILED}
```

Comment:
I find it a bit strange to have yields here and then the return. I wonder, if just for the sake of readability, the except clause would just set a "control failed" variable, and then in the sample loop at the end, there would be a check:
```
# Check samples
for sample in samples:
    result = self._analyze_sample(sample) if not controls_failed else COVID_RESPONSE_FAILED
    yield {"id": sample["id"], "diagnosis_result": result}
```

Reply:
Will fix.